File: app/src/main/java/com/scop/org/minesweeper/generators/solver/Sketch.java
package com.scop.org.minesweeper.generators.solver;

import com.scop.org.minesweeper.elements.Grid;
import com.scop.org.minesweeper.elements.Tile;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class Sketch implements Iterable<Tile>{
	private Grid grid;
	private ArrayList<Integer> numbered;
	private ArrayList<Integer> uncovered;

	public Sketch(Grid grid) {
		this.grid = grid.clone();
		this.numbered = new ArrayList<>();
		this.uncovered = new ArrayList<>();

		for (int i=0; i<getTiles().size(); i++){
			Tile t = getTile(i);
			if (t.isUncovered()){
				uncovered.add(i);
				if (t.getStatus() != Tile.Status.A0) {
					numbered.add(i);
				}
			}
		}
	}

	@Override
	public Iterator<Tile> iterator() {
		return grid.getGrid().iterator();
	}

	public List<Tile> getTiles() {
		return grid.getGrid();
	}

	public Tile getTile(int idx) {
		return getTiles().get(idx);
	}

	public Grid getGrid() {
		return grid;
	}

	public List<Integer> getNumbered(){
		return numbered;
	}

	@SuppressWarnings("unchecked")
	public List<Integer> getNumberedCopy(){
		return (List<Integer>) numbered.clone();
	}

	public List<Integer> getUncovered(){
		return uncovered;
	}

	@SuppressWarnings("unchecked")
	public List<Integer> getUncoveredCopy(){
		return (List<Integer>) uncovered.clone();
	}

	public void addUncovered(Integer i){
		uncovered.add(i);
	}

	public void addNumbered(Integer i){
		numbered.add(i);
		uncovered.add(i);
	}

	public void removeNumbered(Integer i){
		numbered.remove(i);
	}
}

File: app/src/main/java/org/oar/minesweeper/generators/solver/Sketch.kt
package org.oar.minesweeper.generators.solver

import org.oar.minesweeper.elements.Grid
import org.oar.minesweeper.elements.Tile
import org.oar.minesweeper.models.TileStatus

class Sketch(grid: Grid) : Iterable<Tile?> {
    val grid: Grid = grid.clone()

    val numbered: MutableList<Int> = mutableListOf()
    val numberedCopy: MutableList<Int>
        get() = numbered.toMutableList()

    val uncovered: MutableList<Int> = mutableListOf()
    val uncoveredCopy: MutableList<Int>
        get() = uncovered.toMutableList()

    val tiles: List<Tile>
        get() = grid.tiles

    init {
        tiles.indices.forEach { i ->
            val t = getTile(i)
            if (t.isUncovered) {
                uncovered.add(i)
                if (t.status !== TileStatus.A0) {
                    numbered.add(i)
                }
            }
        }
    }

    override fun iterator(): MutableIterator<Tile> {
        return grid.tiles.iterator()
    }

    fun getTile(idx: Int): Tile {
        return tiles[idx]
    }

    fun addUncovered(i: Int) {
        uncovered.add(i)
    }

    fun addNumbered(i: Int) {
        numbered.add(i)
        uncovered.add(i)
    }

    fun removeNumbered(i: Int) {
        numbered.remove(i)
    }
}